lib/ecto/embedded.ex

```diff
@@ -64,39 +66,215 @@ defmodule Ecto.Embedded do
   @doc """
   Wraps embedded models in changesets.
   """
-  def change(%Embedded{cardinality: :one}, value) do
-    Ecto.Changeset.change(value)
+  def change(%Embedded{cardinality: :one, embed: mod}, value, current) do
+    do_change(value, current, mod)
+  end
+
+  # We accept nil here to make is easier to mark embeds recursively
+  # for deletion, otherwise we would need to guess what to pass [] or nil
+  def change(%Embedded{cardinality: :many, container: :array, embed: mod},
+             value, current) when is_list(value) or is_nil(value) do
+    {pk, _} = primary_key(mod)
+    current = process_current(current, pk)
+
+    changeset_fun = &do_change(&1, &2, mod)
+    pk_fun = fn
+      %Changeset{model: model} -> Map.fetch(model, pk)
+      model -> Map.fetch(model, pk)
+    end
+    map_changes(value || [], pk_fun, changeset_fun, current, [], true) |> elem(1)
+  end
+
+  defp do_change(value, nil, mod) do
+    fields    = mod.__schema__(:fields)
+    embeds    = mod.__schema__(:embeds)
+    changeset = Changeset.change(value)
+    types     = changeset.types
+
+    changes =
+      Enum.reduce(embeds, Map.take(changeset.model, fields), fn field, acc ->
+        {:embed, embed} = Map.get(types, field)
+        Map.put(acc, field, change(embed, Map.get(acc, field), nil))
+      end)
+      |> Map.merge(changeset.changes)
+
+    %{changeset | changes: changes, action: :insert}
+  end
+
+  defp do_change(nil, current, mod) do
+    # We need to mark all embeds for deletion too
+    changes = mod.__schema__(:embeds) |> Enum.map(&{&1, nil})
+    %{Changeset.change(current, changes) | action: :delete}
+  end
+
+  defp do_change(%Changeset{model: current} = changeset, current, _mod) do
+    %{changeset | action: :update}
+  end
+
+  defp do_change(%Changeset{}, _current, _mod) do
+    raise ArgumentError, "embedded changeset does not change the model already " <>
+      "preset in the parent model"
+  end
+
+  defp do_change(value, current, _mod) do
+    changes = Map.take(value, value.__struct__.__schema__(:fields))
+    %{Changeset.change(current, changes) | action: :update}
+  end
+
+  @doc """
+  Goes though embedded changesets and autogenerates field for insert ones
+  """
+  def autogenerate(%Embedded{cardinality: :one, embed: model}, changeset, id_types) do
```

Comment:
Here is autogenerate via callbacks: 184592c13e8603b29f623958d3738d9d6f73bcaa

It doesn't cover autogenerating ids though. We will have to figure it out, we probably want to do it while we run insert callbacks for embeds.